File: services/chess_service/src/main/java/net/chess_platform/chess_service/chess/piece/behavior/KnightBehavior.java
package net.chess_platform.chess_service.chess.piece.behavior;

import java.util.ArrayList;
import java.util.List;

import net.chess_platform.chess_service.chess.Chessboard;
import net.chess_platform.chess_service.chess.move.Move;
import net.chess_platform.chess_service.chess.move.Position;
import net.chess_platform.chess_service.chess.piece.AbstractPiece;
import net.chess_platform.chess_service.chess.piece.Knight;
import net.chess_platform.chess_service.chess.piece.Piece;
import net.chess_platform.chess_service.chess.piece.Piece.Color;

public class KnightBehavior implements PieceBehavior {
    
    @Override
    public boolean canBeAppliedTo(Piece piece) {
        return piece instanceof Knight;
    }

    @Override
    public List<Move> getMoves(Chessboard board, AbstractPiece.Type piece, Color color, int row, int col) {
        var moveList = new ArrayList<Move>();

        boolean upperSecondRow = row - 2 >= 0;
        boolean upperFirstRow = row - 1 >= 0;
        boolean lowerFirstRow = row + 1 < Chessboard.SIZE;
        boolean lowerSecondRow = row + 2 < Chessboard.SIZE;
        boolean leftSecondCol = col - 2 >= 0;
        boolean leftFirstCol = col - 1 >= 0;
        boolean rightFirstCol = col + 1 < Chessboard.SIZE;
        boolean rightSecondCol = col + 2 < Chessboard.SIZE;

        if (upperSecondRow) {
            if (leftFirstCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row - 2, col - 1),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
            if (rightFirstCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row - 2, col + 1),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
        }

        if (upperFirstRow) {
            if (leftSecondCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row - 1, col - 2),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
            if (rightSecondCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row - 1, col + 2),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
        }

        if (lowerFirstRow) {
            if (leftSecondCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row + 1, col - 2),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
            if (rightSecondCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row + 1, col + 2),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
        }

        if (lowerSecondRow) {
            if (leftFirstCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row + 2, col - 1),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
            if (rightFirstCol) {
                var move = Move.createBasicMove(board, new Position(row, col), new Position(row + 2, col + 1),
                        piece, color);
                if (move != null) {
                    moveList.add(move);
                }
            }
        }

        return moveList;
    }

}

File: app/src/chess/pieces/behavior/KnightBehavior.ts
import { Chessboard } from "../../Chessboard";
import { createBasicMove, Move } from "../../moves/Move";
import { Color } from "../AbstractPiece";
import { Knight } from "../Knight";
import { Piece, PieceType } from "../Piece";
import { PieceBehavior } from "./PieceBahavior";

export class KnightBehavior implements PieceBehavior {

    getMoves(board: Chessboard, piece: PieceType | null, color: Color, row: number, col: number): Move[] {
        const moveList: Move[] = [];

        const upperSecondRow = row - 2 >= 0;
        const upperFirstRow = row - 1 >= 0;
        const lowerFirstRow = row + 1 < Chessboard.SIZE;
        const lowerSecondRow = row + 2 < Chessboard.SIZE;
        const leftSecondCol = col - 2 >= 0;
        const leftFirstCol = col - 1 >= 0;
        const rightFirstCol = col + 1 < Chessboard.SIZE;
        const rightSecondCol = col + 2 < Chessboard.SIZE;

        if (upperSecondRow) {
            if (leftFirstCol) {
                const move = createBasicMove(board, { row, col }, { row: row - 2, col: col - 1 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
            if (rightFirstCol) {
                const move = createBasicMove(board, { row, col }, { row: row - 2, col: col + 1 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
        }

        if (upperFirstRow) {
            if (leftSecondCol) {
                const move = createBasicMove(board, { row, col }, { row: row - 1, col: col - 2 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
            if (rightSecondCol) {
                const move = createBasicMove(board, { row, col }, { row: row - 1, col: col + 2 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
        }

        if (lowerFirstRow) {
            if (leftSecondCol) {
                const move = createBasicMove(board, { row, col }, { row: row + 1, col: col - 2 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
            if (rightSecondCol) {
                const move = createBasicMove(board, { row, col }, { row: row + 1, col: col + 2 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
        }

        if (lowerSecondRow) {
            if (leftFirstCol) {
                const move = createBasicMove(board, { row, col }, { row: row + 2, col: col - 1 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
            if (rightFirstCol) {
                const move = createBasicMove(board, { row, col }, { row: row + 2, col: col + 1 }, piece, color);
                if (move != null) {
                    moveList.push(move);
                }
            }
        }

        return moveList;
    }

    canBeAppliedTo(piece: Piece | null): boolean {
        return piece instanceof Knight;
    }

}